### PR TITLE
handle requires_grad when creating buckets for distributed

### PR DESCRIPTION
**Pull request by ngimel**
Partially addresses #2533 (this issue is also present in 0.3)

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -139,10 +138,16 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
             if bucket_bytes >= bucket_bytes_cap:
                 self.bucket_sizes.append(0)
                 bucket_bytes = 0
-            self.bucket_sizes[-1] += 1
             for p in param_tuple:
-                self.bucket_map[p] = len(self.bucket_sizes) - 1
-            bucket_bytes += p.numel() * p.element_size()
+                add_to_bucket = False
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: Wouldn't it be just simpler to do `if not param_tuple[0].requires_grad: continue` at the top of the loop?

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -139,10 +138,16 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
...
+                if p.requires_grad:
+                    self.bucket_map[p] = len(self.bucket_sizes) - 1
+                    add_to_bucket = True
+            if add_to_bucket:
+                bucket_bytes += p.numel() * p.element_size()
+                self.bucket_sizes[-1] += 1
+        if self.bucket_sizes[-1]==0:
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: If you modify the loop to skip the iteration at the very beginning this will become unnecessary as well.

**Comment by ngimel** [comment]
@apaszke you are right, modified commit accordingly.

File: torch/nn/parallel/distributed.py [hunk 3]
```diff
@@ -136,13 +135,14 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
...
         bucket_bytes = bucket_bytes_cap  # to init the first bucket immediately
         for param_tuple in zip(*map(lambda m: m.parameters(), self._module_copies)):
-            if bucket_bytes >= bucket_bytes_cap:
-                self.bucket_sizes.append(0)
-                bucket_bytes = 0
-            self.bucket_sizes[-1] += 1
-            for p in param_tuple:
-                self.bucket_map[p] = len(self.bucket_sizes) - 1
-            bucket_bytes += p.numel() * p.element_size()
+            if param_tuple[0].requires_grad:
```
Comment (apaszke) [inline torch/nn/parallel/distributed.py]: The code could still be simplified and the whole diff could be two lines (I'd rather not move the logic around like you did below):
```python
if not param_tuple[0].requires_grad:
    continue
```
We should also give it some more thought. We will never check these flags again right now, and we should either make sure that they are never changed, or we will be missing gradient reductions for a subset of parameters which is really bad.

Comment (ngimel) [reply torch/nn/parallel/distributed.py]: What is happening now is also really bad (as detailed in #2533), and it is fundamental deficiency of how distributed data parallel is written now. Now, if you have some parameters that don't require grad, you still include them in the buckets, so all the buckets won't be filled per iteration, and reductions are happening at unpredictable moments in time.

Comment (apaszke) [reply torch/nn/parallel/distributed.py]: It is bad, but it will raise a proper error for sure. This solution can cause some of the parameters to be silently ignored. I agree the previous solution is a bit of a hack, but I don't think that piling more hacks on top of it will help us in the end. We should figure out a way to actually support params with different flags properly.

Comment (ngimel) [reply torch/nn/parallel/distributed.py]: No, it does not raise proper error! Ever since #2464, if a model has some parameters that don't require grads, the reductions would silently not be triggered at the right moments (and will be triggered at the wrong ones). Before #2464, the model that had parameters that did not require grads errored out.

Comment (ngimel) [reply torch/nn/parallel/distributed.py]: I'm all for figuring out proper solutions, but currently parameters that don't require grad are not supported in distributed, and release notes should not say otherwise https://github.com/pytorch/pytorch/releases (distributed and multi-gpu section)

Comment (soumith) [reply torch/nn/parallel/distributed.py]: fixed the release notes.
